audio/pwaudio.c

```diff
@@ -14,26 +14,357 @@
 #define AUDIO_CAP "pipewire"
 #include "audio_int.h"
 
+typedef struct
+{
+    Audiodev *dev;
+    struct pw_thread_loop *thread_loop;
+    struct pw_context *context;
+
+    struct pw_core *core;
+    struct spa_hook core_listener;
+    struct spa_source *timer;
+} pwaudio;
+
+static void G_GNUC_PRINTF (2, 3) qpw_logerr (int err, const char *fmt, ...)
+{
+    va_list ap;
+
+    va_start (ap, fmt);
+    AUD_vlog (AUDIO_CAP, fmt, ap);
+    va_end (ap);
+
+    AUD_log (AUDIO_CAP, "Reason: %s\n", strerror (err));
+}
+
+static void *qpw_audio_init(Audiodev *dev)
+{
+/*    AudiodevPwOptions *popts = &dev->u.pw;
+*/
+    assert(dev->driver == AUDIODEV_DRIVER_PW);
+    return dev;
+}
+
+typedef struct PWVoiceOut {
+    HWVoiceOut  hw;
+    struct pw_stream *stream;
+    pwaudio *g;
+}
+PWVoiceOut;
```

Comment:
Not a big fan of typedefing structs, but either way, the alias could go in the previous line, after the end of the struct declaration.